desktop/desktopentry/expand_exec.go

```diff
@@ -0,0 +1,141 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package desktopentry
+
+import (
+	"bytes"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/snapcore/snapd/strutil/shlex"
+)
+
+func shellQuote(s string) string {
+	return "'" + strings.Replace(s, "'", "'\\''", -1) + "'"
+}
+
+func toFilePath(uri string) (string, error) {
+	u, err := url.Parse(uri)
+	if err != nil {
+		return "", err
+	}
+	if !u.IsAbs() {
+		return "", fmt.Errorf("%q is not absolute", uri)
+	}
+	if u.Scheme != "file" {
+		return "", fmt.Errorf("%q is not a file URI", uri)
+	}
+	return u.Path, nil
+}
+
+func expandMacro(r rune, buf *bytes.Buffer, de *DesktopEntry, uris []string) ([]string, error) {
+	switch r {
+	case 'u':
+		if len(uris) > 0 {
+			buf.WriteString(shellQuote(uris[0]))
+			uris = uris[1:]
+		}
+	case 'U':
+		first := true
+		for _, u := range uris {
+			buf.WriteString(shellQuote(u))
+			if first {
+				buf.WriteRune(' ')
```

Comment:
I think this will add an extra space after last URI if 2+ URIs are passed, but I guess it is not a problem since it will be split on whitespace anyway? `shlex.Split(buf.String())`

Reply:
You're right. I've adjusted the logic and added extra tests to cover more than 2 URLs. You are right that the we could just add spaces anyway, since the spec says the macro can't be part of a larger argument.

> The %F and %U field codes may only be used as an argument on their own.

I don't think either the glib or KDE implementations do that though, so it seems better to match that behaviour.